scipy/interpolate/_bsplines.py

```diff
@@ -1561,3 +1563,432 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
+            denom = (1 - sum(sum(tr)) / n)**2
+        except LinAlgError:
+            # cholesky decomposition cannot be performed
+            raise ValueError('Seems like the problem is ill-posed')            
+        
+        res = numer / denom
+        
+        return res
+
+    n = X.shape[1]
+
+    XtWX = compute_banded_symmetric_XT_W_Y(X, w, X)
+    XtE = compute_banded_symmetric_XT_W_Y(X, w, wE)
+
+    fun = lambda lam: _gcv(lam, X, XtWX, wE, XtE)
+    gcv_est = minimize_scalar(fun, bounds=(0, n), method='Bounded')
+    if gcv_est.success:
+        return gcv_est.x
+    raise ValueError(f"Unable to find minimum of the GCV "
+                     f"function: {gcv_est.message}")
+
+
+def _coeff_of_divided_diff(x):
+    '''
+    Returns the coefficients of the divided difference.
+
+    Parameters
+    ----------
+    x : array, shape (n,)
+        Array which is used for the computation of divided difference.
+
+    Returns
+    -------
+    res : array_like, shape (n,)
+        Coefficients of the divided difference.
+    
+    Notes
+    -----
+    Vector ``x`` should have unique elements, otherwise an error division by
+    zero might be raised.
+
+    No checks are performed.
+
+    '''
+    n = x.shape[0]
+    res = np.zeros(n)
+    for i in range(n):
+        pp = 1.
+        for k in range(n):
+            if k != i:
+                pp *= (x[i] - x[k])
+        res[i] = 1. / pp
+    return res
+
+
+def make_smoothing_spline(x, y, w=None, lam=None):
+    r'''
+    Returns a smoothing cubic spline function using the GCV criteria [1].
+
+    A smoothing spline is found as a solution to the regularized weighted
+    linear regression problem:
+    
+    .. math::
+
+        \sum\limits_{i=1}^n w_i\abs{y_i - f(x_i)}^2 + 
+        \lambda\int\limits_{x_1}^{x_n} (f^{(2)}(u))^2 \diff u
+
+    where :math:`f` is a spline function, :math:`w` is a vector of weights and
+    :math:`\lambda` is a regularization parameter.
+
+    If ``lam`` is None, we use the GCV criteria to find an optimal
+    regularization parameter, otherwise we solve the regularized weighted
+    linear regression problem with given parameter.
+
+    Parameters
+    ----------
+    x : array_like, shape (n,)
+        Abscissas.
+    y : array_like, shape (n,)
+        Ordinates.
+    w : array_like, shape (n,), optional
+        Vector of weights. Default is ``np.ones_like(x)``.
+    lam : float, (:math:`\lambda \geq 0`), optional
+        Regularization parameter. If ``lam`` is None, then it is found from
+        the GCV criteria. Default is None.
+    
+    Returns
+    -------
+    func : a BSpline object.
+        A callable representing a spline in the B-spline basis 
+        as a solution of the problem of smoothing splines using
+        the GCV criteria [1].
+    
+    Notes
+    -----
+    The original implementation was introduced by Woltring in FORTRAN [2], but
+    because of the license issues cannot be used in the SciPy source code.
```

Comment:
Consider rephrasing: this is  a clean-room reimplementation of the code in [2], correct?